godot-core/src/builtin/projection.rs

```diff
@@ -0,0 +1,443 @@
+    /// clipped.
+    ///
+    /// _Godot equivalent: Projection.get_z_near()_
+    pub fn z_near(&self) -> f64 {
+        self.as_inner().get_z_near()
+    }
+
+    /// Returns a Projection that performs the inverse of this Projection's
+    /// projective transformation.
+    ///
+    /// _Godot equivalent: Projection.inverse()_
+    pub fn inverse(self) -> Self {
+        self.glam(|mat| mat.inverse())
+    }
+
+    /// Returns `true` if this Projection performs an orthogonal projection.
+    ///
+    /// _Godot equivalent: Projection.is_orthogonal()_
+    pub fn is_orthogonal(&self) -> bool {
+        self.as_inner().is_orthogonal()
+    }
+
+    /// Returns a Projection with the X and Y values from the given [`Vector2`]
+    /// added to the first and second values of the final column respectively.
+    ///
+    /// _Godot equivalent: Projection.jitter_offseted()_
+    pub fn jitter_offseted(&self, offset: Vector2) -> Self {
+        self.as_inner().jitter_offseted(offset)
+    }
+
+    /// Returns a Projection with the near clipping distance adjusted to be
+    /// `new_znear`.
+    ///
+    /// Note: The original Projection must be a perspective projection.
+    ///
+    /// _Godot equivalent: Projection.perspective_znear_adjusted()_
+    pub fn perspective_znear_adjusted(&self, new_znear: f64) -> Self {
+        self.as_inner().perspective_znear_adjusted(new_znear)
+    }
+
+    #[doc(hidden)]
+    pub(crate) fn as_inner(&self) -> InnerProjection {
+        InnerProjection::from_outer(self)
+    }
+}
+
+impl From<Transform3D> for Projection {
+    fn from(trans: Transform3D) -> Self {
+        trans.glam(Mat4::from)
+    }
+}
+
+impl Default for Projection {
+    fn default() -> Self {
+        Self::IDENTITY
+    }
+}
+
+impl Mul for Projection {
+    type Output = Self;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        self.glam2(&rhs, |a, b| a * b)
+    }
+}
+
+impl Mul<Vector4> for Projection {
+    type Output = Vector4;
+
+    fn mul(self, rhs: Vector4) -> Self::Output {
+        self.glam2(&rhs, |m, v| m * v)
+    }
+}
+
+impl GlamType for Mat4 {
+    type Mapped = Projection;
+
+    fn to_front(&self) -> Self::Mapped {
+        Projection {
+            cols: [
+                self.col(0).to_front(),
+                self.col(1).to_front(),
+                self.col(2).to_front(),
+                self.col(3).to_front(),
+            ],
+        }
+    }
+
+    fn from_front(mapped: &Self::Mapped) -> Self {
+        Self {
+            x_axis: mapped.cols[0].to_glam(),
+            y_axis: mapped.cols[1].to_glam(),
+            z_axis: mapped.cols[2].to_glam(),
+            w_axis: mapped.cols[3].to_glam(),
+        }
+    }
+}
+
+impl GlamConv for Projection {
+    type Glam = Mat4;
+}
+
+impl GodotFfi for Projection {
+    ffi_methods! { type sys::GDExtensionTypePtr = *mut Self; .. }
+}
+
+/// A projections clipping plane.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
```

Comment:
Nitpick, but try to use consistent order of `#[derive]` arguments (everywhere, not just this line).

This is not 100% uniform across the existing code, but maybe we could use the following order (pick only the ones you need)? 
```rs
#[derive(Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd, Debug)]
```

Reply:
from what i can tell the most common order is:
```rs
#[derive(Default, Copy, Clone, Eq, PartialEq, Debug, Ord, PartialOrd, Hash)]
```
I can do what you said, though i dont like either ordering. 
Personally i'd probably prefer:
```rs
#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
```
Especially i think it makes sense to have traits derived in the order they're required, for instance `PartialEq` before `Eq` because `Eq` depends on `PartialEq`.

Reply:
The order `Eq, PartialEq` has the advantage that you see the important traits first. Which was also the rationale of having `Debug` at the end.

You would then also have grouping by:
* Ways to create an object: `Default`, `Copy`, `Clone`
* Equivalence and ordering: `Eq, PartialEq, Hash, Ord, PartialOrd`
* Printing: `Debug`

So maybe:
```rs
#[derive(Default, Copy, Clone, Eq, PartialEq, Hash, Ord, PartialOrd, Debug)]
```

Reply:
To me it just seems strange to have `Eq` before `PartialEq`, for three main reasons: 
1. rust-analyzer suggests the ordering `PartialEq, Eq`

2. The main behavior of `Eq/Ord/Clone` actually comes from the trait they depend on. In order to compare two functions with `==`, it needs to implement `PartialEq`. `Eq` does not enable you to use `==`, it just makes it usable in more contexts that require equality. Similar for the other two.

3. when i read code, especially a single line, it usually is the case that things are ordered in the order they happen. 

like if i do some kind of iterator chain i'd do
```rs
thing.iter().map(something).filter(another_thing).collect()
```
Which here describes the operation happening from left to right.

Now supertrait relationships are something that is ordered, just like iterator operations. For something to have `Eq` it must first have `PartialEq`. And while the order in the code doesn't matter to rust, to me it still makes sense to order things in the order they mentally "happen" anyway.

Ordering things by importance instead of the order in which they "happen" makes sense if you're encapsulating information. Like if im creating a struct relationship like this: 
```rs
pub struct Foo {
    field: Bar
}

struct Bar { ... }
```
It makes sense that `Bar` comes after `Foo`, because it is `Foo` that is most important here, even though it depends on `Bar`. And if we were to do it the other way around, we'd get cluttered and more difficult to read code. However here we have a couple of differences still. for one a struct definition is generally speaking multiple lines of code, whereas a derive is a single word. and also here we do still name `Bar` inside of `Foo` to say that `Foo` depends on `Bar` before we actually do anything else that depends on `Bar`, like implementing a function that uses it.

So my general my point is that things should at least be *named* before something that depends on it is named, as examples:
- type definitions come before type-impls
- constructors usually come before functions that act on an object constructed by the constructor
- constants are declared before they're used
- variables are declared before they're used
- use-statements come before the imported names are used

etc.

But ordering derives like `Eq, PartialEq` to me violates that principle, as `Eq` depends on `PartialEq`.
The times when this principle is in some way violated is usually because doing it the other way creates a lot of clutter and difficulty reading the code. like in the `Foo/Bar` example above. but that's not really the case for `Eq/PartialEq`.

Reply:
> Does RA really consistently suggest the other one, or did it learn from how you tend to list the two?
It has always suggested `PartialX` before `X`:
![Screenshot_20230302_193502](https://user-images.githubusercontent.com/9107253/222520597-d48cecd3-bc98-4ac1-b27f-d792c8a03a9e.png)